zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/FederatedZkClient.java

```diff
@@ -477,10 +477,31 @@ public long getCreationTime(String path) {
     return getZkClient(path).getCreationTime(path);
   }
 
+  /**
+   * Executes ZkMulti on operations that are connected to the same Zk server.
+   * Will throw exception if any operation's server connection is different.
+   * @param ops
+   * @return
+   * @throws IllegalArgumentException
+   */
   @Override
   public List<OpResult> multi(Iterable<Op> ops) {
-    throwUnsupportedOperationException();
-    return null;
+    if (ops == null) {
+      throw new NullPointerException("ops must not be null.");
+    }
+    String opPath = null;
+    String opPathRealm = null;
+    for (Op op : ops) {
```

Comment:
I'll suggest to implement this logic like this 
`realms = ops.stream()
     .map(op -> getZkRealm(op.getPath()))
     .distinct()
     .collect(toList());`
`if (realms.size() == 1) {
....
} else {
....
}`
This might be a cleaner solution from the code aspect, but your implementation maybe able to fail early if there are different realms. Just a thought, you can think about the trade-offs.

Reply:
Thanks for the suggestion! The solution looks cleaner and while my past solution might fail early, the cost of getting the zkrealm is minimum. I modified your implementation slightly, let me know if it looks good!